### PR TITLE
exclude static features from learned option param spaces

### LINKED ISSUE #856: option learning: remove static features from option parameter space

### PR DESCRIPTION
**Pull request by tomsilver**
closes #856 

I'm going to test this on supercloud with stick button. I would expect that the results for the parameterized methods stay the same or improve. if they don't improve, I can think of two possibilities: there's a bug in this PR, or the `_create_absolute_option_param` function is significantly slow, leading to more timeouts.

### CODE REVIEW

File: src/nsrt_learning/option_learning.py [hunk 1]
```diff
@@ -546,3 +566,18 @@ def _create_regressor(self) -> Regressor:
...
+                                  changing_var_to_feat: Dict[Variable,
+                                                             Set[int]],
+                                  var_order: Sequence[Variable],
+                                  var_to_obj: VarToObjSub) -> Array:
+
+    vec = []
+    for v in var_order:
+        obj = var_to_obj[v]
+        obj_vec = state[obj]
+        for idx in sorted(changing_var_to_feat[v]):
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/nsrt_learning/option_learning.py:581]: (i actually would prefer this)

**Review by ronuchit — COMMENTED** [review]
looks reasonable. how are the results?

File: src/nsrt_learning/option_learning.py [hunk 2]
```diff
@@ -298,7 +295,10 @@ def __init__(self,
             params_space = Box(0, 1, (0, ), dtype=np.float32)
         self._operator = operator
         self._regressor = regressor
-        self._changing_parameter_idxs = changing_parameter_idxs
+        self._changing_var_to_feat = changing_var_to_feat
+        self._changing_vars = [
+            v for v in operator.parameters if v in changing_var_to_feat
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/nsrt_learning/option_learning.py:300]: after reading the rest of the PR, coming back here... i think we can do
`changing_parameters = sorted(changing_var_to_feat, key=op.parameters.index)` like you do elsewhere?

[the remark this answers is not in the document]
Comment (ronuchit) [reply src/nsrt_learning/option_learning.py:300]: oh but also, why do we need to compute changing_parameters elsewhere? we can always just use this self._changing_vars right?

**Comment by ronuchit** [comment]
update docstring of the LearnedNeuralParameterizedOption class, esp the part that says
The objects that are changing correspond to changing_parameters; all other
    objects are assumed to have no change in their state.

File: src/nsrt_learning/option_learning.py [hunk 3]
```diff
@@ -546,3 +566,18 @@ def _create_regressor(self) -> Regressor:
...
             grid_num_ticks_per_dim=num_ticks)
+
+
+def _create_absolute_option_param(state: State,
+                                  changing_var_to_feat: Dict[Variable,
+                                                             Set[int]],
+                                  var_order: Sequence[Variable],
+                                  var_to_obj: VarToObjSub) -> Array:
+
+    vec = []
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/nsrt_learning/option_learning.py:577]: no, I didn't notice any slowdown yet

File: src/nsrt_learning/option_learning.py [hunk 4]
```diff
@@ -412,31 +415,32 @@ def learn_option_specs(self, strips_ops: List[STRIPSOperator],
...
-                    absolute_params = final_state.vec(changing_objects)
-                    option_param = absolute_params - initial_state.vec(
-                        changing_objects)
+                    init_param = _create_absolute_option_param(
+                        init_state, changing_var_to_feat, changing_parameters,
+                        var_to_obj)
+                    final_param = _create_absolute_option_param(
+                        final_state, changing_var_to_feat, changing_parameters,
+                        var_to_obj)
+                    option_param = final_param - init_param
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/nsrt_learning/option_learning.py:443]: yeah

File: src/nsrt_learning/option_learning.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/nsrt_learning/option_learning.py]: I tried to address this, lmk what you think

**Comment by tomsilver** [comment]
>looks reasonable. how are the results?

I'll update the PR when they come in, it will be several hours

**Comment by tomsilver** [comment]
oops did not mean to close

File: src/nsrt_learning/option_learning.py [hunk 5]
```diff
@@ -244,9 +244,11 @@ class _LearnedNeuralParameterizedOption(ParameterizedOption):
       below), all concatenated into a 1D vector, and outputs an action.
 
     The continuous parameters of the option are the main thing to note: they
-    correspond to a desired change in state for a subset of the option objects.
-    The objects that are changing correspond to changing_parameters; all other
-    objects are assumed to have no change in their state.
+    correspond to a desired change in state for a subset of the option objects
+    and a subset of those object features. The objects that are changing are
+    changing_var_order, and also the keys of changing_var_to_feat. The feature
```
Comment (ronuchit) [inline src/nsrt_learning/option_learning.py:249]: and also the keys -> which are the same as the keys

File: src/nsrt_learning/option_learning.py [hunk 6]
```diff
@@ -315,10 +316,11 @@ def _precondition_based_initiable(self, state: State, memory: Dict,
             # The memory here is used to store the absolute params, based on
             # the relative params and the object states.
             memory["params"] = params  # store for sanity checking in policy
```
Comment (ronuchit) [inline src/nsrt_learning/option_learning.py:318]: is this sanity check just verifying that the params here are always the same as when passed into the policy()? this is always guaranteed for any option in ANY situation, right?

File: src/nsrt_learning/option_learning.py [hunk 7]
```diff
@@ -329,9 +331,11 @@ def _regressor_based_policy(self, state: State, memory: Dict,
         if self._is_parameterized:
             # Compute the updated relative goal.
             assert np.allclose(params, memory["params"])
-            changing_objs = [objects[i] for i in self._changing_parameter_idxs]
-            relative_goal_vec = memory["absolute_params"] - state.vec(
-                changing_objs)
+            var_to_obj = dict(zip(self._operator.parameters, objects))
+            state_params = _create_absolute_option_param(
+                state, self._changing_var_to_feat, self._changing_var_order,
```
Comment (ronuchit) [inline src/nsrt_learning/option_learning.py:336]: the state here can and will be different than the above one in `initiable()` right? (it'll match for the first timestep, but then there will be many more after that)

File: src/nsrt_learning/option_learning.py [hunk 8]
```diff
@@ -412,31 +416,29 @@ def learn_option_specs(self, strips_ops: List[STRIPSOperator],
...
-            ]
-
+            changing_var_order = sorted(changing_var_to_feat,
+                                        key=op.parameters.index)
             for segment, var_to_obj in datastore:
                 all_objects_in_operator = [
                     var_to_obj[v] for v in op.parameters
                 ]
 
                 # First, determine the absolute goal vector for this segment.
```
Comment (ronuchit) [inline src/nsrt_learning/option_learning.py:431]: is this comment up to date? i don't understand what here is the "absolute goal vector". i feel like this chunk is actually computing the option parameters.

Comment (ronuchit) [reply src/nsrt_learning/option_learning.py:431]: oh, is final_params the absolute goal vector? can we just clarify this comment significantly, to state that we're accomplishing two things here: (1) computing option_param so they can later be stored into the segments, (2) computing final_param which is the absolute goal vector so we can later compute relative goal vectors when we iterate through the segment's states

File: tests/nsrt_learning/test_option_learning.py [hunk 9]
```diff
@@ -182,15 +182,18 @@ def test_learned_neural_parameterized_option():
                                    pick_nsrt.delete_effects,
                                    pick_nsrt.side_predicates)
     # In this example, both of the parameters (block and robot) are changing.
-    changing_parameters = pick_operator.parameters
+    # For simplicity, assume that the first and third features for the
+    # block and the robot are changing.
+    changing_var_to_feat = {p: {0, 2} for p in pick_operator.parameters}
```
Comment (ronuchit) [inline tests/nsrt_learning/test_option_learning.py:187]: type error: should be a list here: [0, 2]

File: src/nsrt_learning/option_learning.py [hunk 10]
```diff
@@ -412,31 +416,29 @@ def learn_option_specs(self, strips_ops: List[STRIPSOperator],
             # We do this for learning efficiency; including all objects would
             # likely work too, but may require more data for the model to
             # realize that those objects' parameters can be ignored.
-            changing_parameter_set = self._get_changing_parameters(datastore)
+            # Furthermore, we only want to include the features of the objects
+            # that exhibit some change in the data.
+            changing_var_to_feat = self._get_changing_features(datastore)
             # Just to avoid confusion, we will insist that the order of the
             # changing parameters is consistent with the order of the original.
```
Comment (ronuchit) [inline src/nsrt_learning/option_learning.py:423]: original -> operator parameters

**Review by ronuchit — APPROVED** [review]
nice work!

File: src/nsrt_learning/option_learning.py (hunk 6 above)
Comment (tomsilver) [reply src/nsrt_learning/option_learning.py:318]: yeah

File: src/nsrt_learning/option_learning.py (hunk 7 above)
Comment (tomsilver) [reply src/nsrt_learning/option_learning.py:336]: right

**Comment by tomsilver** [comment]
this made a very large difference for cover.

## Before
cover_num_train_tasks_perc_solved (1)

## After
cover_num_train_tasks_perc_solved

the results for stick button look comparable to before.

**Pull request merged**